examples/md-flexible/src/domainDecomposition/RegularGridDecomposition.cpp

```diff
@@ -454,3 +406,95 @@ void RegularGridDecomposition::categorizeParticlesIntoLeftAndRightNeighbour(
+                                 &dummyRequest);
+      autopas::AutoPas_MPI_Isend(&oldLocalBoxMax[i], 1, AUTOPAS_MPI_DOUBLE, leftNeighbour, 0, _communicator,
+                                 &dummyRequest);
+    }
+
+    if (_localBoxMax[i] != _globalBoxMax[i]) {
+      autopas::AutoPas_MPI_Isend(&distributedWorkInPlane[i], 1, AUTOPAS_MPI_DOUBLE, rightNeighbour, 0, _communicator,
+                                 &dummyRequest);
+      autopas::AutoPas_MPI_Isend(&oldLocalBoxMin[i], 1, AUTOPAS_MPI_DOUBLE, rightNeighbour, 0, _communicator,
+                                 &dummyRequest);
+    }
+  }
+
+  for (int i = 0; i < _dimensionCount; ++i) {
+    const int leftNeighbour = _neighbourDomainIndices[i * 2];
+    const int rightNeighbour = _neighbourDomainIndices[i * 2 + 1];
+
+    double neighbourPlaneWork, neighbourBoundary, balancedPosition;
+    if (_localBoxMin[i] != _globalBoxMin[i]) {
+      autopas::AutoPas_MPI_Recv(&neighbourPlaneWork, 1, AUTOPAS_MPI_DOUBLE, leftNeighbour, 0, _communicator,
+                                AUTOPAS_MPI_STATUS_IGNORE);
+      autopas::AutoPas_MPI_Recv(&neighbourBoundary, 1, AUTOPAS_MPI_DOUBLE, leftNeighbour, 0, _communicator,
+                                AUTOPAS_MPI_STATUS_IGNORE);
+
+      _localBoxMin[i] =
+          DomainTools::balanceAdjacentDomains(neighbourPlaneWork, distributedWorkInPlane[i], neighbourBoundary,
+                                              oldLocalBoxMax[i], 2 * (_cutoffWidth + _skinWidth));
+    }
+
+    if (_localBoxMax[i] != _globalBoxMax[i]) {
+      double neighbourPlaneWork, neighbourBoundary;
+      autopas::AutoPas_MPI_Recv(&neighbourPlaneWork, 1, AUTOPAS_MPI_DOUBLE, rightNeighbour, 0, _communicator,
+                                AUTOPAS_MPI_STATUS_IGNORE);
+      autopas::AutoPas_MPI_Recv(&neighbourBoundary, 1, AUTOPAS_MPI_DOUBLE, rightNeighbour, 0, _communicator,
+                                AUTOPAS_MPI_STATUS_IGNORE);
+
+      _localBoxMax[i] =
+          DomainTools::balanceAdjacentDomains(distributedWorkInPlane[i], neighbourPlaneWork, oldLocalBoxMin[i],
+                                              neighbourBoundary, 2 * (_cutoffWidth + _skinWidth));
+    }
+  }
+}
+
+#if defined(MD_FLEXIBLE_INCLUDE_ALL)
+void RegularGridDecomposition::balanceWithAllLoadBalancer(const double &work) {
+  std::vector<ALL::Point<double>> domain(2, ALL::Point<double>(3));
+
+  for (int i = 0; i < 3; ++i) {
+    domain[0][i] = _localBoxMin[i];
+    domain[1][i] = _localBoxMax[i];
+  }
+
+  _allLoadBalancer->setVertices(domain);
+  _allLoadBalancer->setWork(work);
+  _allLoadBalancer->balance();
+
+  std::vector<ALL::Point<double>> updatedVertices = _allLoadBalancer->getVertices();
+
+  for (int i = 0; i < 3; ++i) {
+    _localBoxMin[i] = updatedVertices[0][i];
+    _localBoxMax[i] = updatedVertices[1][i];
+  }
+}
+
+void RegularGridDecomposition::deleteAllLoadBalancer() { delete _allLoadBalancer; }
```

Comment:
please use RegularGridDecomposition's destructor for this kind of cleanup ;)

Reply:
or better: use std::unique_ptr

Reply:
I tried to use std::unique_ptr and to use the RegularGridDecompositions Destructor, but it did not work. When destroyed, the ALL load balancer calls "MPI_Comm_free". If I use the RegularGridDecomposition constructor, all will be deleted when the progam exits the scope of the main function, which is after the "MPI_Finalize" has been called. This results in a build error. As far as I know, a unique pointer cannot be deleted manually, but will be deleted when RegularGridDecomposition is deleted. Again this happens after calling MPI_Finalize which leaves us with the same problem as before.

Reply:
many ways to circumvent this.
1. change the scoping of the class containing the unique_ptr, s.t., it is destroyed before the MPI_Finalize call. In this case, the RegularDecomposition class used in the main.cpp file.
2. you can always "delete" the pointer stored by a unique_ptr by calling `.reset(nullptr)`.

In this case, option 1 is better! You could do that by moving everything that isn't MPI_Init/Finalize to a different function or by surrounding it with curly brackets.

Reply:
I ended up using shared_ptr because RegularGridDecomposition is shared between the main function and the Simulation class.